test/unit/azure_rm_test.rb

```diff
@@ -1,31 +1,176 @@
 require_relative '../test_plugin_helper'
+require_relative '../azure_rm_test_helper'
 
 module ForemanAzureRm
   class ForemanAzureRmTest < ActiveSupport::TestCase
-
+    include AzureRmTestHelper
     setup do
-      @test_sdk = mock('test_sdk') #creates an empty object
-      ForemanAzureRm::AzureRm.any_instance.stubs(:sdk).returns(@test_sdk)
-      @azure_cr = FactoryBot.build(:azure_rm)
-      @azure_cr.stubs(:sdk).returns(@test_sdk)
-    end
-
-    test "list all valid regions" do
-      @test_region_sdk = mock('test_region_sdk')
-      @test_sdk.stubs(:list_regions).with('11111111-1111-1111-1111-111111111111').returns(@test_region_sdk)
+      @mock_sdk = mock('mock_sdk') #creates an empty object
+      ForemanAzureRm::AzureRm.any_instance.stubs(:sdk).returns(@mock_sdk)
+      @mock_region_sdk = mock('mock_region_sdk')
+      @mock_sdk.stubs(:list_regions).with('11111111-1111-1111-1111-111111111111').returns(@mock_region_sdk)
       @eastus_region = mock('eastus_region')
       @westus_region = mock('westus_region')
-      @test_region_sdk.stubs(:value).returns([@eastus_region, @westus_region])
+      @mock_region_sdk.stubs(:value).returns([@eastus_region, @westus_region])
       @eastus_region.stubs(:display_name => 'East US', :name => 'eastus')
       @westus_region.stubs(:display_name => 'West US', :name => 'westus')
+
+      @azure_cr = FactoryBot.create(:azure_rm)
+      @azure_cr.stubs(:sdk).returns(@mock_sdk)
+    end
+
+    test "list all valid regions" do
       assert_equal [['East US', 'eastus'], ['West US', 'westus']], @azure_cr.regions
     end
 
     test "list all resource groups" do
-      @test_resource_client = mock('test_resource_client')
-      @test_sdk.stubs(:resource_client).returns(@test_resource_client)
-      @test_sdk.stubs(:rgs).returns(['rg1', 'rg2', 'rg3'])
+      @mock_resource_client = mock('mock_resource_client')
+      @mock_sdk.stubs(:resource_client).returns(@mock_resource_client)
+      @mock_sdk.stubs(:rgs).returns(['rg1', 'rg2', 'rg3'])
       assert ['rg1', 'rg2', 'rg3'], @azure_cr.resource_groups
     end
+
+    context 'sdk access' do
+      setup do
+        mock_sdk_results
+        @azure_cr.key_pair = @azure_cr.setup_key_pair
+      end
+
+      test "create vm with password and without custom data" do
+        @mock_vm.stubs(:disable_password_authentication).returns(false)
+        @mock_os_profile.stubs(:custom_data)
+        @mock_vm_extension = mock('mock_vm_extension')
+        @mock_sdk.expects(:create_or_update_vm_extensions).with() do |actual_rg, vm_name, script_name, actual_ext|
+          actual_rg == "rg1" &&
+          actual_ext.settings["commandToExecute"] == "$echo testpswd123 | sudo -S echo '\"testuser\" ALL=(ALL) NOPASSWD: ALL' > /etc/sudoers.d/waagent"
+        end.returns(@mock_vm_extension)
+        @mock_sdk.expects(:create_or_update_vm).with() do |actual_rg, actual_vm_name, actual_vm_params|
```

Comment:
It's a bit dicey if we should have this `create_or_update_vm` block in every test? I had to add it in every test as to check the count of sshkeys here on line-51 below and similarly in other tests.